src/main/flight/imu.c

```diff
@@ -345,31 +338,35 @@ static void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
     gy *= (0.5f * dt);
     gz *= (0.5f * dt);
 
-    const float qa = q0;
-    const float qb = q1;
-    const float qc = q2;
-    q0 += (-qb * gx - qc * gy - q3 * gz);
-    q1 += (qa * gx + qc * gz - q3 * gy);
-    q2 += (qa * gy - qb * gz + q3 * gx);
-    q3 += (qa * gz + qb * gy - qc * gx);
+    const float qa = q.w;
+    const float qb = q.x;
+    const float qc = q.y;
+    q.w += (-qb * gx - qc * gy - q.z * gz);
+    q.x += (qa * gx + qc * gz - q.z * gy);
+    q.y += (qa * gy - qb * gz + q.z * gx);
+    q.z += (qa * gz + qb * gy - qc * gx);
 
     // Normalise quaternion
-    recipNorm = invSqrt(sq(q0) + sq(q1) + sq(q2) + sq(q3));
-    q0 *= recipNorm;
-    q1 *= recipNorm;
-    q2 *= recipNorm;
-    q3 *= recipNorm;
+    recipNorm = invSqrt(sq(q.w) + sq(q.x) + sq(q.y) + sq(q.z));
+    q.w *= recipNorm;
+    q.x *= recipNorm;
+    q.y *= recipNorm;
+    q.z *= recipNorm;
 
     // Pre-compute rotation matrix from quaternion
     imuComputeRotationMatrix();
 }
 
-STATIC_UNIT_TESTED void imuUpdateEulerAngles(void)
-{
-    /* Compute pitch/roll angles */
-    attitude.values.roll = lrintf(atan2f(rMat[2][1], rMat[2][2]) * (1800.0f / M_PIf));
-    attitude.values.pitch = lrintf(((0.5f * M_PIf) - acosf(-rMat[2][0])) * (1800.0f / M_PIf));
-    attitude.values.yaw = lrintf((-atan2f(rMat[1][0], rMat[0][0]) * (1800.0f / M_PIf) + magneticDeclination));
+STATIC_UNIT_TESTED void imuUpdateEulerAngles(void){
+    if (FLIGHT_MODE(HEADFREE_MODE)) {
+       attitude.values.roll = lrintf(atan2f((+2.0f * (headfree.wx + headfree.yz)), (+1.0f - 2.0f * (headfree.xx + headfree.yy))) * (1800.0f / M_PIf));
+       attitude.values.pitch = lrintf(asinf(+2.0f * (headfree.wy - headfree.xz)) * (1800.0f / M_PIf));
+       attitude.values.yaw = lrintf((-atan2f((+2.0f * (headfree.wz + headfree.xy)), (+1.0f - 2.0f * (headfree.yy + headfree.zz))) * (1800.0f / M_PIf)));
```

Comment:
I think they are.
headfree quaternion is a by dislocation (yaw angle) rotated version of main quaternion q.